Symfony 2.7 Form choice option update

weaverryan opened the pull request:
Hi guys!

This replaces #5755 #5847 and should close #5179 (or at least covers most things).

| Q | A |
| --- | --- |
| Doc fix? | yes |
| New docs? | yes |
| Applies to | 2.7+ |
| Fixed tickets | #5179 |

This is a big change, and I rewrote a lot. If you have time to review, I think it's best to read the published version up platform.sh and check the code for validity. You can see the code in a project that inspired all the examples here: https://github.com/weaverryan/form_choices/tree/finished.

aivus committed:
Fix docs related with choice form type refactor

Changes related with http://symfony.com/blog/new-in-symfony-2-7-choice-form-type-refactorization

althaus committed:
Added basic choice type changes

althaus committed:
Added choice_loader basics

althaus committed:
Added choices_as_values description

althaus committed:
Added choices_as_values description [amend]

althaus committed:
Fixed linebreaking

althaus committed:
Fixed preferred choices example

weaverryan committed:
Merge branch 'patch-1' of https://github.com/aivus/symfony-docs into form-choice-options

* 'patch-1' of https://github.com/aivus/symfony-docs:
  Fix docs related with choice form type refactor

weaverryan committed:
Merge branch 'improve_form_choice_reference' of https://github.com/althaus/symfony-docs into form-choice-options

* 'improve_form_choice_reference' of https://github.com/althaus/symfony-docs:
  Fixed preferred choices example
  Fixed linebreaking
  Added choices_as_values description [amend]
  Added choices_as_values description
  Added choice_loader basics
  Added basic choice type changes

weaverryan committed:
changing option order - choice_name and choice_value seem like edge cases

weaverryan commented:
I got the PR target wrong - will merge into 2.7.

weaverryan committed:
Completely re-working all the choice docs, including screenshots - for 2.7 changes

weaverryan committed:
fixing bad reference

weaverryan committed:
removing scaling on images

weaverryan committed:
fixing wrong image name

althaus commented:
@weaverryan Nice work. Especially working out the examples should be a good help. I planned to get rid off that `Status` stuff as this only lead to confusion being a too constructed example.

Could you talk to @webmozart? I had some trouble understanding the `choice_loader` so I just took his comment from the code about partially loading stuff. In addition deprecating the `choice_list`, but still using it, is strange.

althaus commented:
Oh... one important thing I missed: It's important to state the usage of `choice_value` to get around the BC(*) that the form options get rendered with integer based values instead of the key value. This breaks any JS frontend code relying on the keys.

* This is still a real BC in 2.7 in my eyes and I'm not alone. :-1:

javiereguiluz commented on reference/forms/types/choice.rst:
```diff
@@ -85,19 +162,86 @@ choices
+        'choices_as_values' => true,
+    ));
+
+In Symfony 3.0, ``choices_as_values`` option will go away, but it's behavior will
```
Since Symfony 3.0 will be released in 4 weeks, we should not use the future tense to make this article future-proof:

```
In Symfony 3.0, ``choices_as_values`` option isn't defined, but it's behavior is always used:
```

javiereguiluz commented on reference/forms/types/entity.rst:
```diff
@@ -171,18 +166,17 @@ either be a ``QueryBuilder`` object or a Closure. If using a Closure,
-~~~~~~~~~~~
+**type**: ``string`` | ``Doctrine\Common\Persistence\ObjectManager`` **default**: the default entity manager
+
+If specified, the specified entity manager will be used to load the choices
```
Weird "specified" redudancy: `If specified, the specified entity manager ...`

javiereguiluz commented on reference/forms/types/options/group_by.rst.inc:
```diff
@@ -0,0 +1,43 @@
+        },
+    ));
+
+This groups the dates that are within 3 days into "Soon" and everyting else into
```
Typo: `everyting` -> `everything`

weaverryan committed:
tweaks thanks to comments

weaverryan committed:
typo!

weaverryan commented:
@althaus Yes, the `Status` isn't a great example - I removed *most* of them, replaced with `DateTime`.

I added a note about the `choice_value` BC break thing. Yes, this is a BC break, but no it won't be fixed (fixing it now would be *another* BC break). All we can do is document it well - I added a small note, I hope it's obvious enough.

About `choice_loader`, what do you mean by:

> In addition deprecating the choice_list, but still using it, is strange.

Where is it still used? I'm also not totally clear on the use-case of `choice_loader`. From looking at the issue queue, I was surprised by how many people seemed to use `choice_list` - I'm not sure of what it's use-case is either.

Anyways, thanks for the review! We have a lot to do, so I'm merging this now - a bit faster than normal. IF anyone sees any issues, please let me know!

Thanks!

althaus commented:
>Yes, this is a BC break, but no it won't be fixed

I know and it's ok. It's just my rage about the decision been made. ;-)

>to use choice_list - I'm not sure of what it's use-case is either.

We've used ChoiceLists in all the cases where our entities aren't coming from the database, but other sources: Printers coming from CUPS, files from the file system, config options from another backend etc. It simply was a convenient way to map those to `choices`. Those have to be rewritten using the choice_value/label/foo settings or with a ChoiceLoader now.


>>    In addition deprecating the choice_list, but still using it, is strange.

> Where is it still used? 

Sorry, that wasn't well written. What I meant is: Although the *option* `choice_list` itself is deprecated, it's used internally for a lot of stuff (see `Symfony\Component\Form\ChoiceList`). It's just an understanding thingy. It's like only the cool kids are allowed to play with those toys. ;-)

> I'm also not totally clear on the use-case of choice_loader
The ChoiceLoader allows to read the options from some backend service. The whole `entity` form type for example relies on a `DoctrineChoiceLoader`.

OskarStark commented on reference/forms/types/choice.rst:
```diff
@@ -44,37 +50,108 @@ Example Usage
+            ],
+            'choices_as_values' => true,
+            'choice_label' => function($val, $key, $index) {
+                /** @var Category $val */
```
would use ``$category``here instead of ``$val``

OskarStark commented on reference/forms/types/choice.rst:
```diff
@@ -44,37 +50,108 @@ Example Usage
+method::
+
+    $builder
+        ->add('category', 'choice', [
```
you use a line break here...

OskarStark commented on reference/forms/types/choice.rst:
```diff
@@ -85,19 +162,88 @@ choices
-is the item value and the array value is the item's label::
+is the item's label and the array value is the item's value::
+
+    $builder->add('inStock', 'choice', array(
```
...but not here and in some other examples? shouldn't it be consistent?

weaverryan commented:
This has already been merged (but I forgot to close the PR). But @OskarStark I will make your tweaks right now - thx :).

@althaus would you be willing to bootstrap an example using `ChoiceLoader`? I "kind of" understand the use-case, but it still seems like you'd need to inject (for example) the EntityManager into your FormType in order to create the ChoiceLoader that you'd pass to your form field. That seems very much not different than injecting the EntityManager in order to just make the `choices` manually. It seems there's more - related to "partially loading" the choices (as you eluded to). Perhaps the choices aren't fully loaded until/unless you render them? And on submit, it only loads the *one* choice from the database (for example)? That indeed would be interesting.

OskarStark commented:
:+1: @weaverryan thank you

althaus commented:
@weaverryan I can try to take a look at our code and see if I can extract a use case essence of it.

>It seems there's more - related to "partially loading" the choices (as you eluded to).

I took the "partial" stuff from @webmozart's code comments in the `ChoiceLoader` file. Never used it myself.

althaus commented:
@weaverryan When will this be online? Neither 2.7 nor 3.0 branch show any changes. :(

weaverryan commented:
@javiereguiluz is there a problem deploying the docs? This was merged 11 days ago, but @althaus is right - I don't see it on the site.

xabbuh commented:
@althaus It's now there: http://symfony.com/doc/2.8/reference/forms/types/choice.html

althaus commented:
@xabbuh Thanks for taking care. :+1:
